Documentation/coding-guidelines/interop-pinvokes.md

```diff
@@ -0,0 +1,240 @@
+scenario of calling a Windows API that takes a string:
+
+1. Create a SB of the desired capacity (allocates managed capacity) **{1}**
+2. Invoke
+   1. Allocates a native buffer **{2}**  
+   2. Copies the contents if `[In]` _(the default for a `StringBuilder` parameter)_  
+   3. Copies the native buffer into a newly allocated managed array if `[Out]` **{3}** _(also the default for `StringBuilder`)_  
+3. `ToString()` allocates yet another managed array **{4}**
+
+That is *{4}* allocations to get a string out of native code. The best you can do to limit this is to reuse the `StringBuilder`
+in another call but this still only saves *1* allocation. It is much better to use and cache a native buffer- you can then
+get down to just the allocation for the `ToString()` on subsequent calls.
+
+The other issue with `StringBuilder` is that it always copies the return buffer back up to the first null. If the passed back string isn't terminated or is a double-null-terminated string your P/Invoke is incorrect at best.
+
+If you *do* use `StringBuilder` one last gotcha is that the capacity does **not** include a hidden null which is always accounted for in interop. It is pretty common for people to get this wrong as most APIs want the size of the buffer *including* the null. This can result in wasted/unnecessary allocations.
+
+**[USE]** Char arrays from `ArrayPool` or `StringBuffer`.
+
+[Default Marshalling for Strings](https://msdn.microsoft.com/en-us/library/s9ts558h.aspx "MSDN")
+
+> ### Windows Specific
+
+> For `[Out]` strings the CLR will use `CoTaskMemFree` by default to free strings or `SysStringFree` for strings that are marked
+as `UnmanagedType.BSTR`.
+
+> **For most APIs with an output string buffer:**
+
+> The passed in character count must include the null. If the returned value is less than the passed in character count the call has succeeded and the value is the number of characters *without* the trailing null. Otherwise the count is the required size of the buffer *including* the null character.
+
+> - Pass in 5, get 4: The string is 4 characters long with a trailing null.
+> - Pass in 5, get 6: The string is 5 characters long, need a 6 character buffer to hold the null.
+
+> [Windows Data Types for Strings](http://msdn.microsoft.com/en-us/library/dd374131.aspx "MSDN")
+
+Booleans
+--------
+
+Booleans are easy to mess up. The default marshalling for P/Invoke is as the Windows type `BOOL`, where it is a 4 byte value. `BOOLEAN`, however, is a *single* byte. This can lead to hard to track down bugs as half the return value will be discarded, which will only *potentially* change the result. You need to use `[MarshalAs(UnmanagedType.U1)]` or `[MarshalAs(UnmanagedType.I1)]` either should work as `TRUE` is defined as `1` and `FALSE` is defined as `0`. `U1` is technically more correct as it is defined as an `unsigned char`.
```

Comment:
This is what I have now:

Booleans are easy to mess up. The default marshalling for P/Invoke is as the Windows type `BOOL`, where it is a 4 byte value. `BOOLEAN`, however, is a *single* byte. This can lead to hard to track down bugs as half the return value will be discarded, which will only *potentially* change the result. For `BOOLEAN` attributing `bool` with either `[MarshalAs(UnmanagedType.U1)]` or `[MarshalAs(UnmanagedType.I1)]` will work as `TRUE` is defined as `1` and `FALSE` is defined as `0`. `U1` is technically more correct as it is defined as an `unsigned char`.